Dashboards: Repeating with no clone keys #107372

torkelo opened the pull request:
Related to epic: https://github.com/grafana/grafana/issues/107371 

### Current problems (in main) I am trying to solve

 * Repeating rows involve long/complex scene object keys (which currently non determinstic keys)
* Cloning a layout for repeat requires complex code to update panel-ids (keys), disable options (drag/drop)
  * https://github.com/grafana/grafana/blob/main/public/app/features/dashboard-scene/scene/layout-default/DefaultGridLayoutManager.tsx#L420
  * https://github.com/grafana/grafana/blob/main/public/app/features/dashboard-scene/scene/layout-auto-grid/AutoGridLayoutManager.tsx#L229
* Complex code to determine if a scene object is a read only clone, or source spread out in a lot of places. Involve checking the scene object key for clone patterns. 
* Complex code needed in getLegacyPanelId to ignore the the messy key and just get the "source" panel id
  * https://github.com/grafana/scenes/blob/main/packages/scenes/src/components/VizPanel/VizPanel.tsx#L184 
* URL keys for repeated panels are non-deterministic , and dependent on order of variable options (should include the repeated variable value not the index of the value) 

### Solution idea

Skip the clone keys mess and just copy the source without any modification, relying on other means than the key to figure out if an object is a clone. 

* Use new pathId instead of panel state key for identifing / finding a specific panel (https://github.com/grafana/grafana/pull/109514) 
* Use new property `repeatSourceKey` to help identify objects that are clones (instead of parsing key string)  (https://github.com/grafana/scenes/pull/1215) 

Benefits 
* In edit mode, selecting any repeated element will select all of them (as they have the same key). You can edit any repeated element (as it will always edit the source) 
* No more complex code to clone layouts 
* Generate determinstic viewPanel URL keys (with new pathId)

TODO
* [ ] Reviews
* [x] Update unit tests
* [ ] Lots of manual testing (with newDashboardLayouts=false and true)
* [x] Solo route (https://github.com/grafana/grafana/pull/109837)
* [ ] After above PR is merged need to update it to use pathId  
* [ ] dashboard data source change?

torkelo committed:
No clone keys

torkelo committed:
Remove all the clone keys complexity

torkelo committed:
More changes

torkelo committed:
update e2e test

Sergej-Vlasov commented:
Initial UI issues that I noticed that do not relate to known solo panel issues:

scenes: 
1. incorrect identity path for nested repeat (first repeated row always shown)
2. No drag and drop of repeated panels

dynamic dashboards:
1. Incorrect identity path for nested repeat (first repeated row/tab always shown)
2. Cannot select repeated panel in row/tab/custom/auto
3. Changes to non repeated panel does not detect/save in repeated tab

I'm loking into 1. Incorrect identity path for nested repeat (first repeated row/tab always shown)
I think its due change to `cloneLayout` using `this.clone({})` will push fix shortly

torkelo committed:
Progress

torkelo commented:
thanks for testing! 

> I think its due change to cloneLayout using this.clone({}) will push fix shortly

@Sergej-Vlasov , that is correct, that function should not have to do anything

torkelo commented:
@Sergej-Vlasov I just tried nested repeats and the viewPanel pathId contains all the parent variable values, can you share the dashboard? I tried nested rows (that repeat by two different variables)

torkelo commented:
> Cannot select repeated panel in row/tab/custom/auto 

That is by design, you can only select the source item (it selects automatically when selecting a clone)

@Sergej-Vlasov

torkelo committed:
fix auto grid item

Sergej-Vlasov commented:
@torkelo figured out the source of `v` keyboard shortcut issue. It was because we removed `joinCloneKeys(...)` from `this.clone({})` methods that were enriching viz panel keys. Those keys were used in scenes in `SetPanelAttentionEvent({ panelId: model.state.key }))` and now that they are gone this event always references panels in original repeats row/tab.

Before:
repeat in first row: `panel-1-clone-2`
repeat in second row: `d20225f2-e541-48e7-9daf-121f92d63fa0-clone-1/grid-item-1/panel-1-clone-2`

Now:
repeat in first row: `panel-1-clone-2`
repeat in second row: `panel-1-clone-2`

This will get even worse if we remove `getCloneKey()` from repeats in `AutoGridItem` and `DashboardGridItem`.
Not sure if this is the only place where identical viz panel keys becomes a problem.  

generating pathId might be too expensive on each hover, so maybe storing it on VizPanel might be an answer

torkelo commented:
@Sergej-Vlasov think we can memoize it and we should be fine, we could also cache it somewhere global maybe

torkelo committed:
Finally working cannot move panel into clone row

torkelo commented:
@Sergej-Vlasov setPanelAttention is also debounced so I would not be too worried about performance issues

moustafab committed:
Alerting: Use db pagination for prometheus api

This reintroduces database pagination for the Prometheus API for rules.

Follow up for #109558 to use the new pagination format.

Note: This changes the sort-order to use `NamespaceUID` instead of the fully qualified Folder path. This also changes the cursor format to use the `NamespaceUID` instead of the Folder path and to be in a reverisble format.

Sergej-Vlasov committed:
adjust how we find vizpanel for keyboard shortcuts

torkelo committed:
Fix shortcuts

torkelo committed:
Update

torkelo committed:
Fix test

torkelo committed:
fixing tests

torkelo committed:
Merge branch 'viz-panel-pathId' into viz-panel-pathid-no-clone

torkelo committed:
fix lint

hugohaggmark committed:
Chore: defaults localizationForPlugins behavior (#109778)

* Chore: defaults localizationForPlugins behavior

* chore: fixes backend tests

* chore: fix backend tests

* chore: fix some more backend tests

torkelo committed:
fix tests

matyax committed:
Logs Panel: use provided height for the container (#109786)

LogsPanel: use provided height for the container

macabu committed:
Secret/Keepers: Return already exists error from DB when creating keeper (#109782)

gabor committed:
datasources: querier: better error handling (#109789)

tomratcliffe committed:
Folders: Remove OldFolderPicker component (#109791)

ywzheng1 committed:
NewProvisionedFolderForm: allow empty space, uppercase and dash in folder name (#109615)

* NewProvisionedFolderForm: allow empty space, uppercase and dash in folder name

* validate folder name using form validation

dprokop committed:
Dashboard Migrations v17: Convert minSpan to maxPerRow in panels (#109592)

* Migration to be verified: v17 Convert minSpan to maxPerRow in panels

* Address review comments: use Go idiom for variable naming and sort.Ints for sorting

fastfrwrd committed:
Table: Tooltip from Field (#109428)

* Table: Tooltip by Field

* add colorization support

* more progress on customizing the tooltip based on cell customization

* configurable as an option

* tooltip triggered from a chip in the corner, pinning tooltip

* i18n

* use enum where appropriate

* correctly orient the tooltip to the cell - but at what cost

* clean up some console.logs

* e2e test for the tooltip

* fix global click stuff

* remove console.log

* refine the style of the caret

* caret placement on same-side as alignment to avoid collision with inspect hover

* some updates from self-review

* increase hit target of tooltip caret

* fix width and height auto-sizing especially for dynamic height

* reorganize to pre-calc the per-field stuff

* use linear gradient for triangle

* update e2e to reflect current behavior, that clicking caret multiple times doesn't toggle pinning

* clean up event handlers a bit

* restore test for toggle click

* alright, re-remove the toggle case

* cursor pointer

* remove optional root from Popover for now

* remove this ridiculous autogenerated file

* update some of the text

* kill the cellRefMatrix

* remove unused import

* extract a util for the predicateByName part

* skip the intermediary step for getCellColors

janthoe committed:
chore(rbac): Register settings resources (#109742)

CollinFingar committed:
QueryLibrary: Added new way to trigger events w/ context (#109637)

pstibrany committed:
chore: Close bleve indexes created in temp dir. (#109799)

* Close indexes created in temp dir.

marcsanmi committed:
Datasource/grafana-pyroscope: Add rate aggregation for cumulative profiles (#108546)

Sergej-Vlasov committed:
AutoGridItemRenderer: Hide repeats when dragging to fix select bug (#109798)

hide repeated panels when dragging

torkelo committed:
fix e2e

Sergej-Vlasov committed:
adjust how outliine children are collected for auto and custom grids

tomratcliffe committed:
Chore: Skip more GHA jobs if backend files are not changed (#109803)

yuri-tceretian committed:
Alerting: Fix fromPrometheusConfig (#109807)

matyax committed:
New Logs Panel: Performance and display updates and fixes (#109672)

* LogLine: don't run overflow check with unwrapped logs

* InfiniteScroll: don't listen to scroll events without scroll

* InfiniteScroll: minimize use of element.scrollHeight

* LogLineDetails: constrain to 95vw

* LogList: process logs while streaming

* Remove unused prop

* InfiniteScroll: update test

* Update test

* Inline Log Details: read max width from context

* Remove console

fastfrwrd committed:
Table: Get rid of ContextMenu on right click (#109825)

* Table: Get rid of ContextMenu on right click

* fix a scroll thing for expanded table, git rid of some conditional accessors

* remove context menu test

* remove context menu test

* i18n

* fix lint issue

grafana-pr-automation[bot] committed:
I18n: Download translations from Crowdin (#109834)

New Crowdin translations by GitHub Action

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

hugohaggmark committed:
BackendSrv: incorporates validatePath in existing RxJs pipe (#107672)

* backendSrv: incorporates validatePath in existing RxJs pipe

* chore: updates after PR feedback

Clarity-89 committed:
CommandPalette: Use fuzzySearch util from grafana/data (#108884)

* CommandPalette: Use fuzzySearch util from grafana/data

* Comment

torkelo committed:
Merge branch 'main' of github.com:grafana/grafana into viz-panel-pathid

torkelo committed:
Merge branch 'viz-panel-pathId' into viz-panel-pathid-no-clone

torkelo committed:
Merge branch 'viz-panel-pathid-no-clone' of github.com:grafana/grafana into viz-panel-pathid-no-clone

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=109614) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=109614) before we can accept your contribution.<br/>**16** out of **17** committers have signed the CLA.<br/><br/>:white_check_mark: Sergej-Vlasov<br/>:white_check_mark: torkelo<br/>:white_check_mark: hugohaggmark<br/>:white_check_mark: gabor<br/>:white_check_mark: macabu<br/>:white_check_mark: moustafab<br/>:white_check_mark: matyax<br/>:white_check_mark: fastfrwrd<br/>:white_check_mark: janthoe<br/>:white_check_mark: ywzheng1<br/>:white_check_mark: dprokop<br/>:white_check_mark: tomratcliffe<br/>:white_check_mark: pstibrany<br/>:white_check_mark: Clarity-89<br/>:white_check_mark: CollinFingar<br/>:white_check_mark: marcsanmi<br/>:x: grafana-pr-automation[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=109614) it.</sub>
